### PR TITLE
Add crypto provider wrappers

### PR DESCRIPTION
**Pull request by steveharter**
Port additional provider wrappers to System.Security.Cryptography.Algorithms and  System.Security.Cryptography.Csp

Address issues:
https://github.com/dotnet/corefx/issues/9981 Add Crypto provider wrappers
https://github.com/dotnet/corefx/issues/12323 Port additional System.Security.Cryptography.RandomNumberGenerator members

Includes misc refactoring to RandomNumberGenerator files (to split into "Implementation" classes under the Interop folder).

Did not add 'sealed' modifier to any classes that did not have them in desktop; at one point this was discussed as an option to help with wrapping (it wasn't necessary).

@bartonjs please review.

### CODE REVIEW

File: src/System.Security.Cryptography.Algorithms/tests/RijndaelTests.cs [hunk 1]
```diff
@@ -11,7 +11,7 @@ namespace System.Security.Cryptography.Encryption.Rijndael.Tests
     using Rijndael = System.Security.Cryptography.Rijndael;
 
     /// <summary>
-    /// Since the Rijndael classes wrap Aes, we only test minimally here since Aes has the full test suite.
+    /// Since Rijndael and RijndaelManaged classes wrap Aes, we only test minimally here.
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/RijndaelTests.cs]: Well, Rijndael doesn't actually wrap AES, only RijndaelManaged does...

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RNGCryptoServiceProvider.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            _impl = Create();
+        }
+
+        public override void GetBytes(byte[] data) => _impl.GetBytes(data);
+
+        // GetBytes(data, offset, count) not forwarded here because the base just calls GetBytes(data)
+
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RNGCryptoServiceProvider.cs]: We should probably move this implementation to the opaque type so that `RandomNumberGenerator.Create().GetNonZeroBytes(buf)` works as well on core as it does on desktop.

Since RNGCryptoServiceProvider is sealed the overload doesn't need to stay here.

Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RNGCryptoServiceProvider.cs]: Done

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// See the LICENSE file in the project root for more information.
+
+namespace System.Security.Cryptography
+{
+    internal sealed partial class RandomNumberGeneratorImplementation : RandomNumberGenerator
+    {
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Actually, since you call GetBytes next, and GetBytes does this validation, it shouldn't be needed here at all.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                if (data[i] != 0)
+                {
+                    data[indexOfFirst0Byte++] = data[i];
+                }
+            }
+
+            while (indexOfFirst0Byte < data.Length)
+            {
+                // This should be more than enough to fill the rest in one iteration
+                byte[] tmp = new byte[2 * (data.Length - indexOfFirst0Byte)];
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Rather than allocating more space than we need, how about adding a private GetBytes(byte[], int) overload, and have that be the one we define in each partial, e.g.
```C#
private void GetBytes(byte[] data, int dataLength)
{
    Debug.Assert(data.Length > 0);
    if (!Interop.Crypto.GetRandomBytes(data, dataLength))
    {
        throw Interop.Crypto.CreateOpenSslCryptographicException();	
    }
}
```
Then put the primary overload in the main partial, e.g.
```C#
public sealed override void GetBytes(byte[] data)
{
    ValidateGetBytesArgs(data);
    if (data.Length > 0)
    {
        GetBytes(data, data.Length);
    }
}
```
and then here doing a single allocation outside of the loop, of the right initial length, and refilling it if need be, e.g.
```C#
int bytesRemaining = data.Length - indexOfFirst0Byte;
if (bytesRemaining > 0)
{
    var tmp = new byte[bytesRemaining];
    while (bytesRemaining > 0)
    {
        GetBytes(tmp, bytesRemaining);
        for (int i = 0; i < bytesRemaining; i++)
        {
            if (tmp[i] != 0)
            {
                data[indexOfFirst0Byte++] = tmp[i];
                if (indexOfFirst0Byte >= data.Length) break;
            }
        }
        bytesRemaining = data.Length - indexOfFirst0Byte;
    }
}
```
or something along those lines.

Actually, do we even need to allocate at all?  What if the whole method were instead like:
```C#
private static void GetNonZeroBytes(byte[] data)
{
    ValidateGetBytesArgs(data);
    if (data.Length > 0)
    {
        int offset = 0;
        while (offset < data.Length)
        {
            // Fill the remaining portion of the array with random bytes.
            GetBytes(data, offset, data.Length - offset);

            // Find the first zero in the remaining portion.
            int indexOfFirst0Byte = data.Length;
            for (int i = offset; i < data.Length; i++)
            {
                if (data[i] == 0)
                {
                    indexOfFirst0Byte = i;
                    break;
                }
            }

            // If there were any zeros, shift down all non-zeros,
            // leaving all of the zeros at the end.
            for (int i = indexOfFirst0Byte; i < data.Length; i++)
            {
                if (data[i] != 0)
                {
                    data[indexOfFirst0Byte++] = data[i];
                }
            }

            // Go again on the space filled with zeros, if any.
            offset = indexOfFirst0Byte;
        }
    }
}
```
This would necessitate changing that other overload of GetBytes to be ```GetBytes(byte[] data, int offset, int length)```, and it would necessitate changing the P/Invokes to accept a ```byte*``` rather than a ```byte[]```.

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs [hunk 5]
```diff
@@ -24,20 +20,30 @@ public void Dispose()
...
+            if (offset < 0) throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (offset + count > data.Length) throw new ArgumentException(SR.Argument_InvalidOffLen);
+
+            if (count > 0)
             {
-                throw new ArgumentNullException(nameof(data));
+                byte[] tempData = new byte[count];
+                GetBytes(tempData);
+                Array.Copy(tempData, 0, data, offset, count);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: Ouch.  Per my previous comment, we should change the interop signatures to accept ```byte*``` and put the primary logic in ```GetBytes(byte[] data, int offset, int count)``` rather than in ```GetBytes(byte[] data)```, so that we don't have to allocate and copy here.

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/SHA1Managed.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            return;
+        }
+
+        protected sealed override void Dispose(bool disposing)
+        {
+            _hashProvider.Dispose(disposing);
+            base.Dispose(disposing);
+        }
+
+        private readonly HashProvider _hashProvider;
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/SHA1Managed.cs]: Nit: please move the fields up to the top of the classes

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs [hunk 7]
```diff
@@ -133,6 +133,22 @@ public static void ConcurrentAccess()
...
+        public static void GetNonZeroBytes()
+        {
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
+            {
+                Assert.Throws<ArgumentNullException>("data", () => rng.GetNonZeroBytes(null));
+
+                // Array should not have any zeros
+                byte[] rand = new byte[65536];
+                rng.GetNonZeroBytes(rand);
+                Assert.Equal(-1, Array.FindIndex(rand, r => (r == 0)));
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Instead of the FindIndex call, it could be ```Array.IndexOf(rand, (byte)0)```

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 3 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: I removed it. Thanks. Fyi that line and the remaining were ported directly from netfx, except for some misc formatting (not saying that we can't improve on it however)

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 4 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: The original code may be trying to minimize the number of writes to the output (i.e. because it fills in 'data' first with the bytes instead of filling a temp buffer first and then copying to data) and also perhaps minimizing the number of allocations + calls to get random data if there's some padding to fill (i.e. 2*bytesremaining) because it is more efficient to do it once than do it 2+ more times in smaller chunks. For example, with a large buffer of 65536 it would on a 'perfect' distribution have to allocate\get random numbers 3 times if 2* was not used (65536[]->256[] -> 1[]) instead of 2 times (65536[]->512[]).

Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: > For example, with a large buffer of 65536 it would on a 'perfect' distribution have to allocate\get random numbers 3 times if 2* was not used (65536[]->256[] -> 1[]) instead of 2 times (65536[]->512[]).

But with my latter solution, there are zero allocations regardless of the buffer length and number of zeros.

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs (hunk 7 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Done. Thanks

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs [hunk 8]
```diff
@@ -177,5 +193,24 @@ private static void DifferentParallel(int arraySize)
...
+                Assert.Throws<NotImplementedException>(() => rng.GetNonZeroBytes(null));
+            }
+        }
+#endif
+
+        private class RandomNumberGeneratorMininal : RandomNumberGenerator
+        {
+            public override void GetBytes(byte[] data)
+            {
+                throw new NotImplementedException();
```
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Good point. Just removed that throw altogether

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs (hunk 5 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: Working on this; note again the original code does this. But yes avoiding the alloc here is goodness

Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: Yeah, I realize most of this is just from desktop.  It'd be good to do better, though. Thanks!

Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: I can remove the alloc here, but (virtual) GetBytes(data, offset, count) must call (abstract) GetBytes(data) per the existing semantics, so we can't put the primary logic in GetBytes(data, offfset, count)

Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: > (virtual) GetBytes(data, offset, count) must call (abstract) GetBytes(data) per the existing semantics

Ugh. Ok.  Then we should override it in our concrete, internal implementations to do the right thing, and we should add a comment here explaining why it has to remain bad.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 4 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Agree, just pointing out perhaps some of the original intent. In your solution, there are zero allocations however perhaps an extra call to get random numbers for large buffers, which does have some overhead although probably not as much as extra allocs.

Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: I was going to say "there's almost no usage of this method", because, well, there shouldn't be.  But it is stupidly popular.  So maybe it is worth the complication.

And `GetBytes(byte[], offset, length)` seems to be described in ApiCatalog as part of netstandard2.0, so that'd be where we end up anyways after taking that change.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs [hunk 9]
```diff
@@ -9,19 +9,16 @@ internal static partial class Interop
...
-                return GetRandomBytes(pBuf, num);
-            }
+            return CryptoNative_GetRandomBytes(pbBuffer, count);
         }
 
-        [DllImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_GetRandomBytes")]
+        [DllImport(Libraries.CryptoNative)]
         [return: MarshalAs(UnmanagedType.Bool)]
-        internal static unsafe extern bool GetRandomBytes(byte* buf, int num);
+        internal static unsafe extern bool CryptoNative_GetRandomBytes(byte* buf, int num);
```
Comment (bartonjs) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: Change to private?

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.OSX.cs [hunk 10]
```diff
@@ -2,17 +2,18 @@
...
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+
 namespace System.Security.Cryptography
 {
     partial class RandomNumberGeneratorImplementation
     {
-        public sealed override void GetBytes(byte[] data)
+        unsafe private void GetBytes(byte* pbBuffer, int count)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.OSX.cs]: Style: Visibility modifier always comes first, so this should be `private unsafe`

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Unix.cs [hunk 11]
```diff
@@ -2,19 +2,20 @@
...
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+
 namespace System.Security.Cryptography
 {
     partial class RandomNumberGeneratorImplementation
     {
-        public sealed override void GetBytes(byte[] data)
+        unsafe private void GetBytes(byte* pbBuffer, int count)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Unix.cs]: `private unsafe`

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Windows.cs [hunk 12]
```diff
@@ -2,19 +2,20 @@
...
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
+
 namespace System.Security.Cryptography
 {
     partial class RandomNumberGeneratorImplementation
     {
-        public sealed override void GetBytes(byte[] data)
+        unsafe private void GetBytes(byte* pbBuffer, int count)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.Windows.cs]: `private unsafe`

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 13]
```diff
@@ -2,57 +2,82 @@
...
+            GetBytesInternal(data, 0, data.Length);
+        }
 
-            GetBytes(data);
+        public override void GetBytes(byte[] data, int offset, int count)
+        {
+            if (data == null) throw new ArgumentNullException(nameof(data));
+            if (offset < 0) throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNum);
+            if (offset + count > data.Length) throw new ArgumentException(SR.Argument_InvalidOffLen);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Rewrite to `if (count > data.Length - offset)` to avoid integer overflow.  (`offset` has already been validated to be non-negative, so the subtraction is safe)

Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: This also has an off-by-one error in it.  If `count != 0` then `count` must be less than `data.Length - offset`.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 14]
```diff
@@ -2,57 +2,82 @@
...
+                        break;
+                    }
                 }
-            }
 
-            for (int i = indexOfFirst0Byte; i < data.Length; i++)
-            {
-                if (data[i] != 0)
+                // If there were any zeros, shift down all non-zeros,
+                // leaving all of the zeros at the end.
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: It doesn't swap the values, so { 3, 0, 8, 0, 192, 11 } turns into { 3, 8, 192, 11, 192, 11 }, then the last two values get overwritten with the next call to GetBytesInternal.

```C#
// If there were any zeros, shift down all non-zero values, and request new bytes.
//   offset = 0
//   filled data as { 3, 0, 8, 0, 192, 11 }
//   indexOfFirst0Byte = 1
//   { 3, 0, 8, 0, 192, 11 } (indexOfFirst0Byte = 1)
//   { 3, 8, 8, 0, 192, 11 } (indexOfFirst0Byte = 2)
//   { 3, 8, 192, 0, 192, 11 } (indexOfFirst0Byte = 3)
//   { 3, 8, 192, 11, 192, 11 } (indexOfFirst0Byte = 4)
//   offset = 4.
//   The last two bytes will be re-generated on the next pass.
```

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 15]
```diff
@@ -2,57 +2,82 @@
...
+                for (int i = indexOfFirst0Byte; i < data.Length; i++)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: this can safely be `i = indexOfFirst0Byte + 1`.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 16]
```diff
@@ -2,57 +2,82 @@
...
-            {
-                // This should be more than enough to fill the rest in one iteration
-                byte[] tmp = new byte[2 * (data.Length - indexOfFirst0Byte)];
-                GetBytes(tmp);
+        private void GetBytesInternal(byte[] data, int offset, int count)
+        {
+            Debug.Assert(data != null);
+            Debug.Assert(offset >= 0);
+            Debug.Assert(count >= 0);
+            Debug.Assert(offset + count <= data.Length);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Fix the asserts per above.

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs [hunk 17]
```diff
@@ -2,57 +2,82 @@
...
 
-                for (int i = 0; i < tmp.Length; i++)
+            if (count > 0)
+            {
+                unsafe
                 {
-                    if (tmp[i] != 0)
+                    fixed (byte* pb = &data[offset])
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: I've always see us use the + version.  Is this happy?

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs [hunk 18]
```diff
@@ -144,7 +144,48 @@ public static void GetNonZeroBytes()
...
+            }
+        }
+
+        [Fact]
+        public static void GetBytes_Offset_ZeroLength()
+        {
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
+            {
+                byte[] rand = new byte[1] { 1 };
+                rng.GetBytes(rand, 0, 0);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Add a rng.GetBytes(rand, rand.Length, 0) case to prove that one out.

File: src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs (hunk 9 above)
Comment (steveharter) [inline src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs]: Sure

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 13 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Done

Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Not sure where you were implying the off-by-one (before or rewritten). The rewritten is equivalent to the before version except for subtracting `offset` on both sides of the > sign.

 If `count != 0` then `count` must be <= `data.Length - offset` (not <)

i.e. this randomizes the last byte and `1 <= 3 - 2`
````
            var rng = RandomNumberGenerator.Create();
            byte[] ran = new byte[3];
            rng.GetBytes(ran, 2, 1);
````

If `count==0` then there's this weirdness (allowed)
````
            rng.GetBytes(new byte[0], 0, 0);
            rng.GetBytes(new byte[1], 1, 0);
            rng.GetBytes(new byte[2], 2, 0);
````

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 16 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Done

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 14 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Comment changes

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs (hunk 18 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Done (no-op)

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 17 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: I have a test for it :) ... I found this on msdn (see &arr[5] example) https://msdn.microsoft.com/en-us/library/f58wzh21.aspx although for bytes just doing a + would be pretty standard I guess

File: src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs (hunk 13 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs]: Yeah, I got confused somewhere.  I think I missed the equal in inequality inversion.

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs [hunk 19]
```diff
@@ -133,6 +133,68 @@ public static void ConcurrentAccess()
...
+
+                rng.GetBytes(rand, 0, 0);
+                Assert.Equal(1, rand[0]);
+
+                // Having an offset of Length is allowed if count is 0
+                rng.GetBytes(rand, rand.Length, 0);
+                Assert.Equal(1, rand[0]);
+            }
+
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Why does the empty array get a new RNG?

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs [hunk 20]
```diff
@@ -133,6 +133,68 @@ public static void ConcurrentAccess()
...
+                rng.GetBytes(rand, 100, 100);
+
+                // Array should not have been touched outside of 100-199
+                Assert.Equal(0, rand[0]);
+                Assert.Equal(99, Array.IndexOf<byte>(rand, 1));
+                Assert.Equal(0, rand[200]);
+                Assert.Equal(299, Array.IndexOf<byte>(rand, 1, 200));
+
+                // Ensure 100-200 has some random bytes; not likely to ever fail here (256^5)
+                Assert.True(rand[100] + rand[125] + rand[150] + rand[175] + rand[199] > 0);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Something like `rand.Skip(100).Take(100).Sum(b => (int)b)` only false negatives at 256^100.

For best results, make offset and count not be the same, since technically we could be interpreting them backwards.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RNGCryptoServiceProvider.cs [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+            if (cspParams != null)
+                throw new PlatformNotSupportedException();
+
+            // This class wraps RandomNumberGenerator.Create() from Algorithms assembly
+            _impl = Create();
+        }
+
+        public override void GetBytes(byte[] data) => _impl.GetBytes(data);
+
+        // GetBytes(data, offset, count) not forwarded here because the base just calls GetBytes(data)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RNGCryptoServiceProvider.cs]: The base does, but the _impl doesn't... if we forward this to the _impl we get the lower-alloc gains.

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SHA1CryptoServiceProvider.cs [hunk 22]
```diff
@@ -0,0 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ComponentModel;
+
+namespace System.Security.Cryptography
+{
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class SHA1CryptoServiceProvider : SHA1
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SHA1CryptoServiceProvider.cs]: Are SHA256, SHA384, and SHA512-CryptoServiceProvider not included?

Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SHA1CryptoServiceProvider.cs]: (They might've been left out if Xamarin didn't have them, I'm just double-checking)

File: src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs (hunk 20 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/tests/RandomNumberGeneratorTests.cs]: Thanks good advice. Made 100-200-100 the new sizes and did the Sum()

File: src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SHA1CryptoServiceProvider.cs (hunk 22 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SHA1CryptoServiceProvider.cs]: I confirmed they are not in the 2.0 list; did not verify Xamerin

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs [hunk 23]
```diff
@@ -24,20 +20,36 @@ public void Dispose()
...
-            if (data == null)
+            VerifyGetBytes(data, offset, count);
+            if (count > 0)
             {
-                throw new ArgumentNullException(nameof(data));
+                // For compat we can't avoid an alloc here since we must call GetBytes(data).
+                // However RandomNumberGeneratorImplementation avoids extra allocs.
+                byte[] tempData = new byte[count];
+                GetBytes(tempData);
+                Array.Copy(tempData, 0, data, offset, count);
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: I understand we can't avoid the allocation/copy in the general case.  But could we for the specific case where the offset/count are for the whole array?
```C#
VerifyGetBytes(data, offset, count);
if (count > 0)
{
    if (offset == 0 && count == data.Length)
    {
        GetBytes(data);
    }
    else
    {
        var tempData = new byte[count];
        GetBytes(tempData);
        Buffer.BlockCopy(tempData, 0, data, offset, count);
    }
}
```

File: src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.cs [hunk 24]
```diff
@@ -7,6 +7,28 @@
 
 namespace System.Security.Cryptography
 {
+#if netcoreapp11
```
Comment (stephentoub) [inline src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.cs]: Does this mean these types show up in the netcore contracts but not in the netstandard ones?

File: src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs (hunk 23 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/RandomNumberGenerator.cs]: Yes I can do that. Thanks

File: src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.cs (hunk 24 above)
Comment (steveharter) [inline src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.cs]: Currently there is an issue with inconsistent contracts that I tried to work around with netcoreapp11 (exists in 462 contract but not in implementation causing genfacades to fail)... I'm working with @joperezr on that before I can commit this PR. It looks like just the AesCryptoServiceProvider type has the issue, so at a minimum that attribute will only exist on that type until we can get the contracts stuff fixed

**Comment by steveharter** [comment]
Test Innerloop Windows_NT Debug Build and Test (System.IO.Tests.DirectoryNotFoundExceptionInteropTests)

**Comment by steveharter** [comment]
Test Innerloop Windows_NT Release Build and Test (System.IO.Tests.DirectoryNotFoundExceptionInteropTests)

**Comment by steveharter** [comment]
Test Innerloop Windows_NT Debug Build and Test (System.IO.Tests.FileNotFoundExceptionInteropTests.From_HR)

File: src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.csproj [hunk 25]
```diff
@@ -3,13 +3,20 @@
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
     <OutputType>Library</OutputType>
-    <NuGetTargetMoniker>.NETStandard,Version=v1.7</NuGetTargetMoniker>
+    <NuGetTargetMoniker Condition="'$(TargetGroup)' == ''">.NETStandard,Version=v1.7</NuGetTargetMoniker>
+    <DefineConstants Condition="'$(TargetGroup)' == 'netcoreapp1.1'">$(DefineConstants);netcoreapp11</DefineConstants>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.csproj]: Is this change needed now that the .builds change was undone? (And nothing uses the netcoreapp11 \#define?)

Comment (steveharter) [inline src/System.Security.Cryptography.Csp/ref/System.Security.Cryptography.Csp.csproj]: PR https://github.com/dotnet/corefx/pull/13081
